### PR TITLE
Add system fallback in GetUserProfileDir

### PR DESCRIPTION
**Pull request by am11**
### Problem
Most of the popular tools, such as `git`, work in environments regardless `HOME` (or `USERPROFILE` on Windows) environment variable is set.

In .NET 7, we are trying to make the user experience slightly more friendly in minimal / single-user OS installations, especially containers and sandboxes, where HOME variable might not be set in the default configuration. This will allow users to fall into pit of success.

In .NET 6 and earlier versions, this is what the $HOME-less user gets:

```sh
# unset HOME
dotnet --help
The user's home directory could not be determined. Set the 'DOTNET_CLI_HOME' environment variable to specify the directory to use.
````

In .NET 7 preview 5 (daily build, 7.0.100-preview.5.22267.11), with https://github.com/dotnet/runtime/commit/cef245d28ae8d50f3bf0ad7adedc461cd45b085c and https://github.com/dotnet/sdk/commit/e0456aaf3d9df757b76af6f055eedcca3b51e4da in, most commands successfully execute with `unset HOME` and do not run into the error shown above.

However, `dotnet new` command fails as follow:

```sh
# unset HOME
dotnet7 new console -n test42

Unhandled exception: System.NotSupportedException: HOME or USERPROFILE environment variable is not defined, the environment is not supported
   at Microsoft.TemplateEngine.Edge.DefaultPathInfo.GetUserProfileDir(IEnvironment environment)
   at Microsoft.TemplateEngine.Edge.DefaultPathInfo..ctor(IEngineEnvironmentSettings engineEnvironmentSettings, String settingsLocation)
   at Microsoft.TemplateEngine.Edge.EngineEnvironmentSettings..ctor(ITemplateEngineHost host, Boolean virtualizeSettings, String settingsLocation, IEnvironment environment, IComponentManager componentManager, IPathInfo pathInfo)
   at Microsoft.TemplateEngine.Cli.Commands.BaseCommand.CreateEnvironmentSettings(GlobalArgs args, ParseResult parseResult)
   at Microsoft.TemplateEngine.Cli.Commands.BaseCommand`1.InvokeAsync(InvocationContext context)
   at Microsoft.TemplateEngine.Cli.Commands.BaseCommand`1.Invoke(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.DotNet.Cli.Parser.<>c__DisplayClass16_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass12_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__18_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass8_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

### Solution
This PR ports CLI's implementation  from https://github.com/dotnet/sdk/blob/830b7c45020c004f1958b8e74d3b21d4a45dbf22/src/Common/CliFolderPathCalculatorCore.cs#L30 to DefaultPathInfo, which works like this:
* try to obtain value of `DOTNET_CLI_HOME` environment variable
* when the above fails, try to get it from `HOME` on Unix or `USERPROFILE` on Windows.
* when the above fails, fallback to `System.Environment.GetFolderPath(SpecialFolder.UserProfile)` which, in addition to honoring the environment variables, fallback to calling underlying operating system's API to obtain the path.
* in SDK, we return null if all of the above fails, but here I have kept the existing exception.

Note that it is very unlikely that GetUserProfileDir() will throw an exception, i.e. `System.Environment.GetFolderPath(SpecialFolder.UserProfile)` does not return null or empty string on officially supported platforms today. I am not convinced that there is a value to wire `GetFolderPath` through `IEnvironment` for UserProfile, as this functionality is robustly provided by the corelib (from both coreclr and mono runtimes alike). If you think there is a potential extensibility scenario where it could bring value, please let me know in the comments.

Thanks.

### Checks:
- [x] Added unit tests
- [ ] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Edge/DefaultPathInfo.cs [hunk 1]
```diff
@@ -97,9 +97,25 @@ public sealed class DefaultPathInfo : IPathInfo
...
         private static string GetUserProfileDir (IEnvironment environment)
         {
-            bool isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
-            return environment.GetEnvironmentVariable(isWindows ? "USERPROFILE" : "HOME")
-                ?? throw new NotSupportedException("HOME or USERPROFILE environment variable is not defined, the environment is not supported");
+            const string DotnetHomeVariableName = "DOTNET_CLI_HOME";
+            string platformHomeVariableName = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "USERPROFILE" : "HOME";
+            string? home = environment.GetEnvironmentVariable(DotnetHomeVariableName);
+
+            if (string.IsNullOrEmpty(home))
```
Comment (am11) [inline src/Microsoft.TemplateEngine.Edge/DefaultPathInfo.cs:104]:
```suggestion

            string? home = environment.GetEnvironmentVariable(DotnetHomeVariableName);
            if (string.IsNullOrEmpty(home))
```

**Review by JanKrivanek — CHANGES_REQUESTED** [review]
Please extract the calls to `Environment` to IEnvironment (it can be abstracted to a single property - e.g. `UserProfileFolder`) and add a test variant for fallback to it.

Other than that it looks like a great improvement - thank you!

**Comment by am11** [comment]
Thanks for your quick review, @JanKrivanek. I have addressed the feedback.

**Review by JanKrivanek — APPROVED** [review]
Thank you Adeel @am11. Looks good to go to me.
Vlada @vlada-shubina  - do you want to have one more look before merging?

File: src/Microsoft.TemplateEngine.Edge/DefaultPathInfo.cs [hunk 2]
```diff
@@ -97,9 +97,24 @@ public sealed class DefaultPathInfo : IPathInfo
 
         private static string GetUserProfileDir (IEnvironment environment)
         {
-            bool isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
-            return environment.GetEnvironmentVariable(isWindows ? "USERPROFILE" : "HOME")
-                ?? throw new NotSupportedException("HOME or USERPROFILE environment variable is not defined, the environment is not supported");
+            const string DotnetHomeVariableName = "DOTNET_CLI_HOME";
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/DefaultPathInfo.cs:100]: Unfortunately we cannot use `DOTNET_CLI_HOME` here as it is specific to dotnet CLI, and `Edge` is used by other parties too.

The way forward may be to add another parameter for `DefaultPathInfo` that will specify the overriden user folder location. If it's not set, the default (previous version) may be used. `TemplateEngine.CLI` should check if `DOTNET_CLI_HOME` is set and if so, when creating `EngineIEnvironmentSettings` here https://github.com/dotnet/templating/blob/2851334289e12782b5c9b0264301a9c9bfc0af28/src/Microsoft.TemplateEngine.Cli/Commands/BaseCommand.cs#L50 pass the `DefaultPathInfo` with that path set.

Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Edge/DefaultPathInfo.cs:100]: Or do completely separate `IDefaultPathInfo` implementation for `TemplateEngine.CLI`.
Unsealing this class and allow overriding the properties might be another option too to reuse the code for other properties.

**Review by vlada-shubina — CHANGES_REQUESTED** [review]
We cannot use `DOTNET_CLI_HOME` in `TemplateEngine.Edge`. Please see the comments inline.

File: src/Microsoft.TemplateEngine.Edge/DefaultPathInfo.cs (hunk 2 above)
Comment (am11) [reply src/Microsoft.TemplateEngine.Edge/DefaultPathInfo.cs]: Thanks. I didn't realize the other use-cases of Edge engine. I have added a flag in `EngineEnvironmentSettings`'s ctor which indicates whether or not `DOTNET_CLI_HOME` should be probed. A new test case in included which ensures it is working as expected.

Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Edge/DefaultPathInfo.cs]: We should avoid using `DOTNET_CLI_HOME` at all in `Edge`.

Please do evaluation of `DOTNET_CLI_HOME` in `TemplateEngine.CLI` and pass resolved user folder to `DefaultPathInfo` (via new parameter) or inherit from `DefaultPathInfo`  in `TemplateEngine.CLI` or do own implementation of `IDefaultPathInfo`  in `TemplateEngine.CLI`.

The only change in Edge should be `DefaultPathInfo` accepting custom path for user folder directory or its unsealing.

Comment (am11) [reply src/Microsoft.TemplateEngine.Edge/DefaultPathInfo.cs]: Thanks, I have modified the implementation so `Edge` does not use `DOTNET_CLI_HOME`.

File: test/Microsoft.TemplateEngine.Edge.UnitTests/PathInfoTests.cs [hunk 3]
```diff
@@ -14,6 +14,63 @@ namespace Microsoft.TemplateEngine.Edge.UnitTests
...
+        [Theory]
+        [InlineData (false, false, false)]
+        [InlineData (false, false, true)]
+        [InlineData (false, true, false)]
+        [InlineData (false, true, true)]
+        [InlineData (true, false, false)]
+        [InlineData (true, false, true)]
+        [InlineData (true, true, false)]
+        [InlineData (true, true, true)]
+        public void UserProfileEnvironment(bool useHome, bool useDotnetCliHome, bool useUserProfilePath)
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Edge.UnitTests/PathInfoTests.cs:26]: this test might not be relevant anymore. Instead please consider unit test in `Cli.UnitTests` or integration test in `dotnet-new3.IntegrationTests`. 

Here is an example of integration test, which modifies environment variables before act: https://github.com/dotnet/templating/blob/209d566999ffc0f89deb5948b022c4331715fb1b/test/dotnet-new3.UnitTests/DotnetNewDebugOptions.cs#L103-L115

**Review by vlada-shubina — CHANGES_REQUESTED** [review]
It seems there are some changes left which are no longer needed, could you please review?
Also, please modify `DefaultPathInfo` public constructor instead. 

Thank you.

**Pull request closed without merging**